afpacket/src/tokio_sockets.rs

```diff
@@ -20,11 +35,287 @@ impl AsyncBoundSocket {
         self.sock.get_mut().set_promiscuous(p)
     }
 
-    pub async fn send(&mut self, frame: &[u8]) -> io::Result<usize> {
-        self.sock.write(frame).await
+    pub fn split(self) -> (SendHalf, RecvHalf) {
+        let inner = Arc::new(Inner {
+            locked: AtomicBool::new(false),
+            sock: UnsafeCell::new(self),
+        });
+
+        let rx = RecvHalf {
+            inner: inner.clone(),
+        };
+        let tx = SendHalf { inner };
+
+        (tx, rx)
+    }
+
+    pub fn poll_can_rx(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        let ready = Ready::readable();
+        ready!(self.sock.poll_read_ready(cx, ready))?;
+        Poll::Ready(Ok(()))
+    }
+
+    pub fn poll_can_tx(&mut self, cx: &mut Context<'_>) -> Poll<io::Result<()>> {
+        ready!(self.sock.poll_write_ready(cx))?;
+        Poll::Ready(Ok(()))
+    }
+
+    pub fn clear_can_tx(&mut self, cx: &mut Context<'_>) -> io::Result<()> {
+        self.sock.clear_write_ready(cx)
+    }
+
+    pub fn clear_can_rx(&mut self, cx: &mut Context<'_>) -> io::Result<()> {
+        let ready = Ready::readable();
+        self.sock.clear_read_ready(cx, ready)
+    }
+
+    pub fn poll_send(&mut self, cx: &mut Context<'_>, frame: &[u8]) -> Poll<io::Result<usize>> {
+        ready!(self.poll_can_tx(cx))?;
+        match self.sock.get_mut().send(frame) {
+            Ok(count) => Poll::Ready(Ok(count)),
+            Err(ref e) if e.kind() == io::ErrorKind::WouldBlock => {
+                self.clear_can_tx(cx)?;
+                Poll::Pending
+            }
+            Err(e) => Poll::Ready(Err(e)),
+        }
+    }
+
+    pub fn poll_recv(
+        &mut self,
+        cx: &mut Context<'_>,
+        frame: &mut [u8],
```

Comment:
Are there size requirements for frame?  What happens if the data to write would overflow the buffer? Panic? Should be documented.